Modify GetNextWorkRequired to set Target Limit correctly

CaveSpectre11 opened the pull request:
GetNextWorkRequired() was not setting the work required properly on the first Proof of Stake block.  The code section to cap the work at the bnTargetLimit wasn't called on that first block, but rather on the second Proof of Stake block.

The end result is that the first PoS block uses the difficulty from the PoW, rather than capping it at the target limit; with the second PoS block capping it at the target limit:

`uint256 bnTargetLimit = (~uint256(0) >> 24);`

Creating a difficulty problem should the final PoW difficulty be greater than bnTargetLimit.

---
This code has been tested on an altcoin; and I can provide debug output demonstrating the shift of the TargetLimit to 00000FFFF the block after the first PoS block should it be desired.

Mrs-X commented:
Hmh, I have to check this more closely, but on a quick look  `LoadExternalBlockFile()` would fail because `ProcessNewBlock()` would fail because `CheckWork()`would fail because `GetNextWorkRequired()` would return a different `nBitsRequired`, thus `if (block.nBits != nBitsRequired)` would apply for the first PoW block.

CaveSpectre11 commented:
Ok, my priority of getting a dev environment to test PIVX directly has increased.

Yes, that does appear to be the case on inspection (my initial issue was on the shift from PoW to PoS; but since this seems to be a potential reindex failure, it may not be innocuous to PIVX).

CaveSpectre11 commented:
Running the test overnight.  Have a fix coded up that I will also validate before pushing.

CaveSpectre11 commented:
@Mrs-X you are correct.  Please see with the change in main.cpp

From my debugging build:

```
2019-06-21 13:00:56 GetNextWorkRequired - pure:
        bnNew:          000000000532ea8748d8748d8748d8748d8748d8748d8748d8748d8748d8748d
2019-06-21 13:00:56 GetNextWorkRequired:
        bnNew:          000000000532ea8748d8748d8748d8748d8748d8748d8748d8748d8748d8748d
        bnTargetLimit:  000000ffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
2019-06-21 13:00:56 CheckWork(): nbits=470117804, required=470102762
2019-06-21 13:00:56 CheckWork(): Accepting special block
2019-06-21 13:00:56 UpdateTip: new best=1c9121bf9329a6234bfd1ea2d91515f19cd96990725265253f4b164283ade5dd  height=259201 version=3  log2_work=55.779856  tx=367159  date=2016-08-17 02:40:14 progress=0.077175  cache=6335
2019-06-21 13:00:56 ProcessNewBlock : ACCEPTED Block 259201 in 72 milliseconds with size=959
```

CaveSpectre11 commented:
> Concept ACK.
> Might make sense to define variables somewhere to store those values for `nTime` and `nBits`, instead of using magic numbers.

I considered that; but I opted against it specifically because of how unique that check is.  Originally I was just checking for the block; but by checking nTime and nBits; that code should not pass if it's ported somewhere other than PIVX.   This way it's obscure, and requires whomever ports the fix into their code base to understand exactly what they're doing, and exactly where they need to do it.  If they have more than one block; it won't be obscured.   If I was to define a variable, I would be compelled to define a vector and have it loop through the vector so it's a more portable).

But that's MHO; and I don't feel strongly about the opinion.  I will pull it out into chainparams and define them if you wish.

random-zebra commented:
I think those are valid arguments so, as far as I am concerned, this can remain as is.
Let's wait for other reviewers.

Warrows commented:
I don't understand why you'd need a vector :thinking: In chainparams you could define two constants like `NEXTWORK_BUG_BLOCKTIME = 1471401614` and `NEXTWORK_BUG_NBITS = 0x1c056dac` and use them in the test. This way you explicit what those numbers are without making the code much more complicated.
I think that's the best way to go about it.

CaveSpectre11 commented:
> I don't understand why you'd need a vector 🤔 

I'm going to try my best to lay out my position without too many tangents; but I can't make any promises :)

One thread that always runs in my coding style is reuse.  It's everywhere in my current day job (basically 7 different forks of the code base I'm working on).  This applies greatly to the whole crypto industry, as well.  When I have something localized in a module, or a function, I'm more likely to opt for the quick route.  When it extrapolates into more of a general or removed use, is where the reuse becomes more relevant.   For example; 950 started with ValidateURL being local to the budget.cpp module.  When it was suggested it be moved to utilities; the URL length became a relevant magic number; and thus I added it as a parameter to the function.  So other callers from other uses may have different requirements for the URL length.  Arguably, the hard coded "64" default in the header declaration could be made into a constant that's defined yet elsewhere.  However, going so far as to add the vectors to allow the URL content checks to be changed by the caller was a bit overkill, however the code is written with that in mind; it would be an easy mod to pass in different search vectors into that routine should it be determined necessary.

So how this all applies to the subject at hand.  I chose to use nbits and time, instead of block number; so that anyone forking pivx doesn't have a hole on that block; when they hit the block.  It now has to hit the exact nbits at the exact time; which is already past.  So the likelihood of another coins block being accepted is slim to none; and it's in the past as well so it shouldn't matter anyway.

For Pivx; it was only one block affected.  For others, it could be none, or many.   That's when the "right" way to code that would be a vector that can cover 1 to many problem blocks.  Yes, once again, a little overkill here.  But I also argue that putting an exception to skip over a consensus problem into chainparams is too far disconnected from the problem to make it readable.   **_If_** there was any reason that it would ever be relevant anywhere else in the code... then absolutely, a constant would be warranted.

Now, here's the parallel for Pivx specifically. For #814 to be handled properly, it needs to ignore the overmint in over a dozen blocks.  Since it doesn't need to be anywhere else except ConnectBlock I would debate that it still belongs locally defined in main.cpp rather than chainparams.  In the fork that I addressed the same issue as 814; it was only one effected block (and thus... coded with a magic number like this PR).  For Pivx, however; and many forks, there is going to be a various number of blocks that need to be accounted for; and thus setting the framework of a vector that is populated for whatever a coin's specific instances are... becomes my recommended approach.

So with all that said... it's why I made the choices I did.  I feel it's actually an appropriate use of magic numbers, and well contained in the location where it applies.  

But, like most code style debates; I only casually date my code... I don't marry it, so if you guys agree "change it", your wish is my command.

random-zebra commented:
For what is worth, I'm in line with these points:

> But I also argue that putting an exception to skip over a consensus problem into chainparams is too far disconnected from the problem to make it readable.

> I feel it's actually an appropriate use of magic numbers, and well contained in the location where it applies.

Warrows commented:
Alright, I get your point and mostly agree. However, I still think magic numbers should be explicitly defined every time it's possible. So even if it's two lines above in main.cpp I'd like to see two constants defined for these numbers.

CaveSpectre11 commented:
>I still think magic numbers should be explicitly defined every time it's possible. So even if it's two lines above in main.cpp I'd like to see two constants defined for these numbers.

Compromise accepted.   All set.

Warrows commented on src/main.cpp:
```diff
@@ -4509,8 +4509,17 @@ bool CheckWork(const CBlock block, CBlockIndex* const pindexPrev)
+    if (block.nBits != nBitsRequired) {
+        // Pivx Specific reference to the block with the wrong threshold was used.
+        static const int64_t PivxBadBlockTime=1471401614;
+        static const unsigned int PivxBadBlocknBits=0x1c056dac;
```
nit: missing spaces around `=` signs.

CaveSpectre11 replied:
taken care of

CaveSpectre11 commented:
> Not a big fan special processing for single blocks/transactions/whatever, 

Agreed.  If the error case wasn't there already, I would have considered letting it be; But this way there's only extra processing when the block is headed down the failure path.

CaveSpectre11 commented:
> TestNet sync will obviously halt at block 201.

Excellent point.  Is there an explorer for regtest so I can get the info on block 251?

Fuzzbawls commented:
Regtest is a non-persistent chain that is initialized from genesis on demand. Testnet has a BE though: https://testnet.pivx.link

CaveSpectre11 commented:
> Regtest is a non-persistent chain that is initialized from genesis on demand. Testnet has a BE though: https://testnet.pivx.link

Ok; regtest won't apply.  testnet I had already pulled.

CaveSpectre11 commented:
Set up for both testnet and mainnet to be covered; put stubs for regtest so regtest won't falter over the code if it ever ends up in it.  I'll check travis and rebase in the morning.

random-zebra commented on src/chainparams.cpp:
```diff
@@ -376,6 +380,8 @@ class CRegTestParams : public CTestNetParams
         bnProofOfWorkLimit = ~uint256(0) >> 1;
         nLastPOWBlock = 250;
+        nPivxBadBlockTime = 1489001494;  // Regtest shouldn't apply, so we will just use testnet settings since they shouldn't match
+        nPivxBadBlocknBits = 0x1e0a20bd; // Regtest shouldn't apply, so we will just use testnet settings since they shouldn't match
```
Regtest is started on demand (with the latest code) and destroyed when the test is over. So it will never need any kind of escaping for bad blocks.

Furthermore `CRegTestParams` is a subclass of `CTestNetParams` so we have called the parent constructor (and set those values for `nPivxBadBlockTime` and `nPivxBadBlocknBits`) already here.

CaveSpectre11 replied:
> Regtest is started on demand (with the latest code) and destroyed when the test is over. So it will never need any kind of escaping for bad blocks.

However, if there was a bad nBit during regtest; it would go into the code path that checks if it's because it's the bad block.  That's what I was trying to cover... (how to ensure the code path to return true wasn't entered on regtest) but as you further explain; it's covered by the inheritance.

CaveSpectre11 replied:
fixed up; and cleaned up my rebase mess too

random-zebra commented:
Merging...
